examples/tabular_q.py

```diff
@@ -0,0 +1,206 @@
+        "-instsimplify",
+        "-jump-threading",
+        "-loop-reduce",
+        "-loop-rotate",
+        "-loop-versioning",
+        "-mem2reg",
+        "-newgvn",
+        "-reg2mem",
+        "-simplifycfg",
+        "-sroa",
+    ],
+    "A list of action names to explore from.",
+)
+flags.DEFINE_float("discount", 1.0, "The discount factor.")
+flags.DEFINE_list(
+    "features_indices",
+    [19, 22, 51],
+    "Indices of Alphaphase features that are used to construct a state",
+)
+flags.DEFINE_float("learning_rate", 0.1, "learning rate of the q-learning.")
+flags.DEFINE_integer("episodes", 5000, "number of episodes used to learn.")
+flags.DEFINE_integer(
+    "log_every", 50, "number of episode interval where progress is reported."
+)
+flags.DEFINE_float("epsilon", 0.2, "Epsilon rate of exploration. ")
+flags.DEFINE_integer("episode_length", 5, "The number of steps in each episode.")
+FLAGS = flags.FLAGS
+
+
+class StateActionTuple(NamedTuple):
+    """An state action tuple used as q-table keys"""
+
+    autophase0: int
+    autophase1: int
+    autophase2: int
+    cur_step: int
+    action_index: int
+
+
+def make_q_table_key(autophase_feature, action, step):
+    """Create a hashable Q-table key.
+
+    For tabular learning we will be constructing a Q-table which maps a
+    (state, action) pair to an expected (remaining) reward. The purpose of this
+    function is to convert the (state, action) properties into a hashable tuple
+    that can be used as a key for a Q-table dictionary.
+
+    In the CompilerGym setup, encoding the true state the program is not obvious,
+    and this solution turns to use the observations from Autophase features instead.
+    The default arguments handpicked 3 indices from the Autophase feature that
+    appear to change a lot during optimization.
+
+    In addition, the current step in the episode is added to the state representation
+    as well. In the current fixed-episode-length setup, we need to differentiate
+    reaching a state at different steps, as they can lead to different final rewards,
+    depending on the remaining optimization steps.
+
+    Finally, we add the action index to the key.
+    """
+
+    return StateActionTuple(
+        *autophase_feature[FLAGS.features_indices], step, FLAGS.actions.index(action)
+    )
+
+
+def select_action(q_table, ob, step, epsilon=0.0):
+    qs = [q_table.get(make_q_table_key(ob, act, step), -1) for act in FLAGS.actions]
+    if random.random() < epsilon:
+        return random.choice(FLAGS.actions)
+    max_indices = [i for i, x in enumerate(qs) if x == max(qs)]
+    # Breaking ties at random by selecting any of the indices.
+    return FLAGS.actions[random.choice(max_indices)]
+
+
+def get_max_q_value(q_table, ob, step):
+    max_q = 0
+    for act in FLAGS.actions:
+        hashed = make_q_table_key(ob, act, step)
+        if hashed not in q_table:
+            q_table[hashed] = 0
+        max_q = max(q_table[hashed], max_q)
```

Comment:
You could use the more pythonic, more compact and faster max_q = max(q_table.get(hashed, 0), max_q) instead of inserting 0 in the q_table. You just need to make sure to use it consistently throughout the code: since hashed will not be inserted in the hashtable, subsequent calls to q_table[hashed] may throw unless you replace them with calls to  q_table.get(hashed, 0)

Reply:
Thanks for the input!